src/GameLogic/Level.ts

```diff
@@ -33,4 +33,17 @@ export class Level {
 		this.tilesFloor = new Grid2D<FloorType>(this.width, this.height, FloorType.FloorTile);
 		this.entities = new Entities();
 	}
+
+	public clone(): Level {
+		const clonedLevel = new Level({
+			width: this.width,
+			height: this.height,
+			playerStartX: this.playerStartX,
+			playerStartY: this.playerStartY,
+		});
+		clonedLevel.tilesFloor.setAllByCallback((x, y) => this.tilesFloor.get(x, y));
+		clonedLevel.entities = new Entities(Array.from(this.entities.entities));
```

Comment:
Looking better, but we also need to clone the entities themselves!

Reply:
Yikes. But I think I've got it now